### DISCUSSION TITLE
Safe migration from Sidecars to Ambient mode

### DISCUSSION THREAD

**Discussion by joze144** [istio/istio#53345] [Q&A] [answered]
What are suggested steps to migrate from using Sidecars to Ambient mode on production?

Is my thinking correct?
1. Update Istio with Ambient profile
2. Add label `istio.io/dataplane-mode=ambient` to namespaces. Is it safe to have sidecars still running at this point?
3. Remove label `istio-injection=enabled` from namespaces.
4. Restart deployments

**Comment by ilrudie** [comment, marked as answer]
This is a good start but there are a few additional things I think you must be aware of which may required you to adjust your plans.

1. by default sidecars would not be able to talk to ztunnel and this is controlled at injection time so depending on whether or not you have prepared for your migration in your last sidecar upgrade you may actually need to restart all deployment in order to get injected sidecars that can communicate over HBONE. Depending on which version(s) you are running just enabling HBONE might not be a good idea and upgrading to a newer release first might be a safer approach (you could consider enabling ambient during this upgrade but wait to begin using the ambient dataplane until all your proxies are able to communicate with HBONE.)
3. the gateways also need to be able to use HBONE, so would have similar requirements as mentioned above.
4. You'll need to consider if your current use of Istio would require waypoint proxies though. For instance if you have some policy that says "the beep service account in boop namespace may perform GET operations on my service" then a waypoint would be needed to enforce it because policy containing HTTP attributes cannot be enforced correctly by ztunnel. 

re: 2's question; yup. that is fine. the sidecar label will take precedence so until you remove the sidecar injection and restart your deployments you are effectively still running sidecar Istio but you've covered that in your steps. Having both labels doesn't cause issues that we have seen and is considered a valid state.

**Comment by kirann-hegde** [comment]
Hello @ilrudie @joze144 

I am facing a rather strange issue with Istio v1.23.4.
Here is my scenario:
-Install Istio v1.20.3 in side car mode in the **istio-system** namespace. Tag the revision as : stable.
-Label a namespace named **bookinfo** with `istio.io/rev=stable`. Check the output of the command: `istioctl ps`. I observe that all the pods in this labelled namespace are part of the mesh and are using the 1.20.3 Istiod version. 
Here is the screenshot
![image](https://github.com/user-attachments/assets/fb411529-c624-42f9-9924-b8ec001878e1)

The labels on the namespace can be seen in the screenshot
![image](https://github.com/user-attachments/assets/e9589871-dcba-45f8-b83a-8f0aa74a62b3)



-I now install Istio v1.24.3 using helm charts into the same **istio-system** namespace as earlier. In the helm chart values, I use the values
```
revision: 1-24-3 
revisionTags: 
- canary
```
Istio v1.24.3 gets installed side by side with Istio .
The output of `istioctl tag list` is as expected.
![image](https://github.com/user-attachments/assets/579be0bf-f095-4720-a842-3a5b660e2fbc)

-I now apply the label `istio.io/rev=canary ` on the same namespace as earlier and restart the pods in this namespace.
-I again execute `istioctl ps` and as expected, the pods are now using Istoid 1.24.3.
All well and good so far.

I now want to switch the pods in the namespace to ambient mode. 
Hence, I add the `istio-injection=disabled,istio.io/dataplane-mode=ambient` to the same namespace and restart all the pods in the namespace.
I again execute the command `istioctl ps` and to my surprise, this time, the pods from the namespace are not part of the output. 
![image](https://github.com/user-attachments/assets/d005297f-97c4-4f1c-bd42-a1f2e281b0ca)

The version of istioctl that I am using is
![image](https://github.com/user-attachments/assets/36f49344-6fde-4f89-8fd0-0d1f4051c444)

For some strange reason, Kiali reports that the **bookinfo** namespace is in ambient mode(Kiali might have a bug and might just be working based upon the labels on the namespace)

![image](https://github.com/user-attachments/assets/d00e8ba0-6cc9-4f5c-887f-69cbdf94071f)


![image](https://github.com/user-attachments/assets/a34e6317-6d29-4547-9a91-e508c2ef30b2)


**Questions**

-What am I doing wrong here?

-How can I switch the workloads to start using ambient mode?

-Can I rely on the output of `istioctl ps` or is there a better way to confirm that the namespace is indeed using the ambient mesh?

-Is it recommended to install both versions of Istio into the same istio-system namespace?

As always, any help would be very much appreciated.

**Reply by kirann-hegde** [reply]
Alright, after some more digging, it seems like the output of `istioctl ps ` only contains pods that contain an envoy proxy. Could someone confirm it?

I can see the traffic to my pods being logged in the logs of the ztunnel. Can I assume that ambient mode is indeed enabled for the  pods in the bookinfo namespace?

**Reply by ilrudie** [reply]
`proxy-status` shows envoy proxy information. There is a new `ztunnel-config` subcommand to show ztunnel information.

https://ambientmesh.io/docs/security/verify-mtls/ has more information about verifying the ambient dataplane is enabled.

**Reply by howardjohn** [reply]
`proxy-status` actually shows Ztunnel as well, maybe you were confusing `proxy-config`?

* `ztunnel-config`: introspect a single ztunnel
* `proxy-config`: introspect a single envoy
* `proxy-status`: view connected XDS clients (of any type) from Istiod
